src/shogun/features/RandomKitchenSinksDotFeatures.cpp

```diff
@@ -35,16 +35,12 @@ RandomKitchenSinksDotFeatures::RandomKitchenSinksDotFeatures(
 
 SGMatrix<float64_t> RandomKitchenSinksDotFeatures::generate_random_coefficients()
 {
-	SGVector<float64_t> vec = generate_random_parameter_vector();
-	SGMatrix<float64_t> random_params(vec.vlen, num_samples);
-	for (index_t dim=0; dim<random_params.num_rows; dim++)
-		random_params(dim, 0) = vec[dim];
+	SGMatrix<float64_t> random_params(num_samples + 1, num_samples);
```

Comment:
why has the dimension now changed? the number of rows should be the dimension of the feature space

Reply:
not sure where this `num_samples + 1` comes from?

Reply:
``` generate_random_parameter_vector()``` returns a vec of vlen  ```get_feature_dim_space() + 1 ```. 

```get_feature_dim_space()```  returns num_samples

So in essence, vec.vlen = num_samples + 1.

Reply:
`generate_random_parameter_vector` is a purely virtual method though, so it could return different things in the future. I realise now why there was a single vector generated first, and then the remaining...in order to be sure about the dimensions of the matrix. I did't realise this before. You will have to generate one parameter vector, then allocates the matrix, and then copy it to the first column of the matrix, then the rest.
BTW I am puzzled what this matrix represents actually. Usually one needs to store frequencies (coefficients) that are dim x num_rf, see e.g. [here](https://github.com/karlnapf/kernel_exp_family/blob/master/kernel_exp_family/kernels/kernels.py#L316)

Reply:
actually thinking about all this more. I am not sure how much worth it even is doing these optimizations here, as the design is really not that good (see my email with many suggestions how to improve it)

Reply:
Ok for now let's do this

* add a method that can tell us the number of rows of this matrix, so that we can allocate it without having to generate a random vec or hard-code anything.
* add a way for the generate random vector to write the results to a pre-specified memory block (the column of the matrix)
* then loop over the columns and write the random parameters directly to the matrix column wise

Reply:
```
auto d = .... // figure out the dimension via a method or similar
SGMatrix<float64_t> random_params(d, num_samples);
for (auto i : range(num_samples))
  generate_random_parameter_vector(random_params.get_column(i)); // not sure this exists, but something in these lines will
```

Reply:
> `generate_random_parameter_vector` is a purely virtual method though, so it could return different things in the future. I realise now why there was a single vector generated first, and then the remaining...in order to be sure about the dimensions of the matrix. I did't realise this before. You will have to generate one parameter vector, then allocates the matrix, and then copy it to the first column of the matrix, then the rest.
> BTW I am puzzled what this matrix represents actually. Usually one needs to store frequencies (coefficients) that are dim x num_rf, see e.g. [here](https://github.com/karlnapf/kernel_exp_family/blob/master/kernel_exp_family/kernels/kernels.py#L316). So num_samples+1 x num_samples doesnt make sense to me. Maybe you could investigate that?

I was thinking the same  about why num_samples + 1 is taken, apparently it stores numbers from gaussiann in [0 .... num_samples] and 1 number from uniform distrib in [num_sample+1]. But isn't the number sampled from uniform_distrb to act as an offset ?.

Reply:
Ah yes of course it’s the offset. 
Another thing to refactor... but that’s for later

I suggest you add a method that returns the number of parameters per random feature (use it inside the random vec generation) and then allocate the matrix using that.

Reply:
Storing the offset in the last component means that we cannot do matrix vector products for the projection. That would be another improvement (in tidiness and speed)

Reply:
Okay upon investigation here's what I found.
```get_dim_feature_space()``` returns the dim of feature space of the final RFF transformed feature( K x m ) { k = number of samples, m = number of examples}
However, our ```random_param``` matrix must have dimensions ```dim_feature_space x num_samples```.(n x m) . Here the feature_space is of the **input vector** not of the final vector.

Let's continue this over the mailing list, meanwhile, I'll investigate further.

Reply:
This is all expceted. Here is a python implementation I wrote a while ago that might make it easier to understand [link](https://github.com/karlnapf/kernel_exp_family/blob/master/kernel_exp_family/kernels/kernels.py#L316)